polarCoordinateTraversal.m

```diff
@@ -111,75 +111,134 @@
 end
 
 % Find the intersection times for the ray and the radial shell containing the ray
-% at t_begin.
+% at t_begin; in order to determine if ray intersects grid.
 ray_unit_vector = 1 / sqrt(ray_direction(1)^2 + ray_direction(2)^2)...
     .* [ray_direction(1);  ray_direction(2)]';
 v = dot(ray_circle_vector,ray_unit_vector);
 discr = r^2 - (dot(ray_circle_vector,ray_circle_vector) - v^2);
 d = sqrt(discr);
-p1 = ray_origin + (v-d).*ray_unit_vector;
-p2 = ray_origin + (v+d).*ray_unit_vector;
-t1 = (p1(1)-ray_origin(1))/ray_direction(1);
-t2 = (p2(1)-ray_origin(2))/ray_direction(2);
+pa = ray_origin + (v-d).*ray_unit_vector;
+pb = ray_origin + (v+d).*ray_unit_vector;
+tol = 10^-16;
+if (ray_direction(1) < tol)
+    t1 = (pa(2) - ray_origin(2))/ray_direction(2);
+    t2 = (pb(2) - ray_origin(2))/ray_direction(2);
+else
+    t1 = (pa(1) - ray_origin(1))/ray_direction(1);
+    t2 = (pb(1) - ray_origin(1))/ray_direction(1);
+end
 
 % The ray may not intersect the grid at all. 
 % In particular, if the ray is outside the grid at t_begin.
-if (t1 < t_begin && t2 < t_begin )
+if t1 < t_begin && t2 < t_begin 
+    if verbose
+        fprintf("\nRay does not intersect polar grid for t_begin.")
+    end
+    return;
+end
+
+% It may be a tangent hit
+ if abs(t1 - t2) < tol
     if verbose
-        fprintf("\nRay does not intersect polar grid for t_begin.");
+        fprintf("\nTangent hit.")
     end
     return;
 end
 
 % If there is a ray/shell intersection, then set the radial voxel ID. 
 current_voxel_ID_r = 1 + (circle_max_radius - r)/delta_radius;
 if verbose
-    fprintf('RADIAL HIT INITIALIZED.\n');
+    fprintf('RADIAL HIT INITIALIZED.\n')
 end
 
-% Calculate Voxel ID Theta.
-current_voxel_ID_theta = floor(atan2(ray_start_y - circle_center_y, ray_start_x - circle_center_x) * num_angular_sections / (2 * pi));
+% II. Calculate Voxel ID Theta.
+tol = 10^-16;
+if abs(ray_origin - circle_center) < tol 
+    % If the ray starts at the origin, we need to perturb slightly along its path to find the
+    % correct angular voxel
+    pert_t = 0.1;
+    pert_x = ray_start_x + ray_direction_x * pert_t;
+    pert_y = ray_start_y + ray_direction_y * pert_t;
+    current_voxel_ID_theta = floor(atan2(pert_y - circle_center_y, pert_x - circle_center_x) * num_angular_sections / (2 * pi));
+else
+    current_voxel_ID_theta = floor(atan2(ray_start_y - circle_center_y, ray_start_x - circle_center_x) * num_angular_sections / (2 * pi));
+end
 if current_voxel_ID_theta < 0
     current_voxel_ID_theta = num_angular_sections + current_voxel_ID_theta;
 end
 
-% Pre-calculations for radial_hit. This is necessary to check for intersections with relevant radial neighbors.
```

Comment:
Did you mean to remove this? This was moved to polarCoordinateTraversal in another PR so we didn't calculate it with each radial_hit call.

Reply:
I meant to remove the comment, yes. The precalcs occur on lines 106, 115, 117